### PR TITLE
display different payment options based on user location

### LINKED ISSUE #3711: detect Brave Payments user location and display different dialog content
In order to prevent users who are outside of the USA from attempting to buy/transfer bitcoin using the Coinbase widget, we need to detect their location and change the visible elements of the dialog.

(see dialog spec: #3677)

USA:
image

ROW:
image

### PR DESCRIPTION
**Pull request by jkup**
- [x] Submitted a [ticket](https://github.com/brave/browser-laptop/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- [x] Ran `git rebase -i` to squash commits (if needed).

Fix #3711 

Test Plan:

I'm having a tough time testing this one. I don't see exchangeInfo.exchangeName or exchangeInfo.exchangeURL even when I VPN to another country. @mrose17 what triggers those to come down? How can I test getting them?

Based on that I might need to change the naming of some of my functions but this is the general idea.

### CODE REVIEW

**Comment by mrose17** [comment]
@jkup - sorry for the delay in replying, it's been a little busy this week.

first, make sure that you are seeing `ledgerInfo.countryCode`. it is the 3166-2 alpha code for the country that the browser is operating in. if it's not there, please let me know!

second, if the code is "US", there will not be any `ledgerInfo.exchangInfo.*` properties

third, if the code isn't "US", and there aren't any `ledgerInfo.exchangInfo.*` properties, please let me know what the country code is

thanks!

**Comment by mrose17** [comment]
i would prefer to see it in 0.12.1 ... however, it isn't essential.

@jkup - can you construct a clean branch from master and then generate a new PR?

thanks,

/mtr

**Comment by jkup** [comment]
@mrose17 I updated the PR so it doesn't have conflicts. I think there is a misunderstanding though in what state data I have access to. I don't see this.props.ledgerInfo only this.props.ledgerData and I don't see countrycode on that.

Do I have access to any of this info in preferences.js?

**Comment by mrose17** [comment]
@jkup - please check with @diracdeltas ... in `ledger.js` it is called `ledgerInfo`, perhaps it goes by a different name in the UI.

to verify what `ledger.js` is sending, run the browser like this:

        LEDGER_CLIENT_+DEBUG=true npm start

and look for a series of lines that starts with

        updateLedgerInfo: ...

good luck!

**Comment by diracdeltas** [comment]
@mrose17 i don't see exchangeInfo either when i run with LEDGER_CLIENT_DEBUG=true

```
updateLedgerInfo: {
  "creating": false,
  "created": true,
  "reconcileStamp": 1475887090152,
  "transactions": [],
  "address": "35x1xVzJgFNmyWoWZva1aaCvCgjnchfNvs",
  "balance": "0.0079",
  "unconfirmed": "0.0000",
  "satoshis": 791348,
  "btc": "0.01643142",
  "amount": 10,
  "currency": "USD",
  "paymentURL": "bitcoin:35x1xVzJgFNmyWoWZva1aaCvCgjnchfNvs?amount=0.01643142&label=Bra$e%20Software",
  "buyURL": "https://buy.coinbase.com?crypto_currency=BTC&code=729cbe57-599a-5a99-b131-$28d9da0987d&amount=10&address=35x1xVzJgFNmyWoWZva1aaCvCgjnchfNvs",
  "bravery": {
    "setting": "adFree",
    "fee": {
      "currency": "USD",
      "amount": 10
    }
  },
  "hasBitcoinHandler": false,
  "countryCode": "US",
  "error": null,
  "paymentIMG": ...
```

File: js/about/preferences.js [hunk 1]
```diff
@@ -311,6 +315,20 @@ class BitcoinDashboard extends ImmutableComponent {
...
+  get exchangePanel () {
+    const exchangeInfo = this.props.ledgerData.get('exchangeInfo')
+    return <div className='panel'>
+      <div className='settingsPanelDivider'>
+        <span className='fa fa-credit-card' />
+        <div className='settingsListTitle' data-l10n-id='outsideUSAPayment' />
+      </div>
+      <div className='settingsPanelDivider'>
+        <a target='_blank' className='browserButton primaryButton' href={exchangeInfo.get('exchangeURL')}>
+          {exchangeInfo.get('exchangeURL')}
```
Comment (diracdeltas) [inline js/about/preferences.js]: should be labeled `exchangeInfo.get('exchangeName')

File: js/about/preferences.js [hunk 2]
```diff
@@ -362,7 +380,10 @@ class BitcoinDashboard extends ImmutableComponent {
         : null
       }
       <div className='board'>
-        {this.coinbasePanel}
+        { (this.userInAmerica)
+          ? this.coinbasePanel
+          : this.exchangePanel
```
Comment (diracdeltas) [inline js/about/preferences.js]: should probably fall back to coinbase panel if exchangeUrl is not defined

Comment (jkup) [inline js/about/preferences.js]: how does this approach look?

Comment (diracdeltas) [inline js/about/preferences.js]: i'm not sure, i just think it should show something reasonable if `exchangeUrl` is empty due to a network connection problem or something

Comment (diracdeltas) [inline js/about/preferences.js]: @jkup oops, missed the context of your question. looks fine.

File: js/about/preferences.js [hunk 3]
```diff
@@ -311,6 +315,27 @@ class BitcoinDashboard extends ImmutableComponent {
       </div>
     }
   }
+  get exchangePanel () {
+    const exchangeInfo = this.props.ledgerData.get('exchangeInfo')
+    const url = exchangeInfo.get('exchangeURL')
+    const name = exchangeInfo.get('exchangeName')
```
Comment (diracdeltas) [inline js/about/preferences.js]: this causes the `Add funds` panel to throw an error and not appear at all if `exchangeInfo` is undefined. better to use `this.props.ledgerData.getIn(['exchangeInfo', propertyname])`

File: js/about/preferences.js [hunk 4]
```diff
@@ -311,6 +315,27 @@ class BitcoinDashboard extends ImmutableComponent {
...
+      return this.coinbasePanel
+    } else {
+      return <div className='panel'>
+        <div className='settingsPanelDivider'>
+          <span className='fa fa-credit-card' />
+          <div className='settingsListTitle' data-l10n-id='outsideUSAPayment' />
+        </div>
+        <div className='settingsPanelDivider'>
+          <a target='_blank' className='browserButton primaryButton' href={exchangeInfo.get('exchangeURL')}>
+            {exchangeInfo.get('exchangeName')}
```
Comment (diracdeltas) [inline js/about/preferences.js]: can use `url`, `name` defined above

Comment (diracdeltas) [inline js/about/preferences.js]: label should be something like `Go to [exchangeName]`

**Comment by jkup** [comment]
@diracdeltas TIL about .getIn 👍

File: js/about/preferences.js [hunk 5]
```diff
@@ -311,6 +315,26 @@ class BitcoinDashboard extends ImmutableComponent {
...
+      return this.coinbasePanel
+    } else {
+      return <div className='panel'>
+        <div className='settingsPanelDivider'>
+          <span className='fa fa-credit-card' />
+          <div className='settingsListTitle' data-l10n-id='outsideUSAPayment' />
+        </div>
+        <div className='settingsPanelDivider'>
+          <a target='_blank' className='browserButton primaryButton' href={url}>
+            {name}
```
Comment (diracdeltas) [inline js/about/preferences.js]: github somehow lost my comment but i think this should say `Go to ${name}` instead of just the name

Comment (jkup) [inline js/about/preferences.js]: Nice catch! I totally forgot to add the word 'visit' before the button like @bradleyrichter has in his designs! Fixed.

**Comment by luixxiul** [comment]
From Japan, visit Cubits :+1:

**Comment by luixxiul** [comment]
For QA: test needed by who lives in USA

**Pull request merged**